From 86654f2bc9f4b3de114107008bb98e5f526dfe8c Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Mon, 26 Apr 2021 17:34:14 +0200
Subject: [PATCH 07/10] audioconvert: run lr4 on tagged channels in generic
 case

Mark the LFE channels and run the lowpass filter on them in
the generic case. Generates LFE correctly in 2.1 upmix case.

See #1095
---
 spa/plugins/audioconvert/channelmix-ops-c.c | 4 ++++
 spa/plugins/audioconvert/channelmix-ops.c   | 9 +++++++--
 spa/plugins/audioconvert/channelmix-ops.h   | 1 +
 3 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/spa/plugins/audioconvert/channelmix-ops-c.c b/spa/plugins/audioconvert/channelmix-ops-c.c
index 588cff19..e7359a93 100644
--- a/spa/plugins/audioconvert/channelmix-ops-c.c
+++ b/spa/plugins/audioconvert/channelmix-ops-c.c
@@ -78,6 +78,10 @@ channelmix_f32_n_m_c(struct channelmix *mix, uint32_t n_dst, void * SPA_RESTRICT
 				d[i][n] = sum;
 			}
 		}
+		for (i = 0; i < n_dst; i++) {
+			if (mix->lr4_info[i] > 0)
+				lr4_process(&mix->lr4[i], d[i], n_samples);
+		}
 	}
 }
 
diff --git a/spa/plugins/audioconvert/channelmix-ops.c b/spa/plugins/audioconvert/channelmix-ops.c
index c669c58b..7c98e645 100644
--- a/spa/plugins/audioconvert/channelmix-ops.c
+++ b/spa/plugins/audioconvert/channelmix-ops.c
@@ -422,10 +422,15 @@ done:
 				continue;
 			mix->matrix_orig[ic][jc++] = matrix[i][j];
 			sum += fabs(matrix[i][j]);
-			if (i == _CH(LFE))
-				lr4_set(&mix->lr4[ic], BQ_LOWPASS, mix->lfe_cutoff / mix->freq);
 		}
 		maxsum = SPA_MAX(maxsum, sum);
+		if (i == _CH(LFE)) {
+			spa_log_debug(mix->log, "channel %d is LFE", ic);
+			lr4_set(&mix->lr4[ic], BQ_LOWPASS, mix->lfe_cutoff / mix->freq);
+			mix->lr4_info[ic] = 1;
+		} else {
+			mix->lr4_info[ic] = 0;
+		}
 		ic++;
 	}
 	if (SPA_FLAG_IS_SET(mix->options, CHANNELMIX_OPTION_NORMALIZE) &&
diff --git a/spa/plugins/audioconvert/channelmix-ops.h b/spa/plugins/audioconvert/channelmix-ops.h
index 3dd600e4..8fc15c4f 100644
--- a/spa/plugins/audioconvert/channelmix-ops.h
+++ b/spa/plugins/audioconvert/channelmix-ops.h
@@ -65,6 +65,7 @@ struct channelmix {
 
 	float freq;					/* sample frequency */
 	float lfe_cutoff;				/* in Hz, 0 is disabled */
+	uint32_t lr4_info[SPA_AUDIO_MAX_CHANNELS];
 	struct lr4 lr4[SPA_AUDIO_MAX_CHANNELS];
 
 	void (*process) (struct channelmix *mix, uint32_t n_dst, void * SPA_RESTRICT dst[n_dst],
-- 
2.30.2

